File: python_implementation/heapsort.py
"""Implementació del HeapSort en Python."""
from excepcions import LlistaBuida, NoEsLLista, ValorsNoNumerics

from colorama import init, Fore


init(autoreset=True)


class HeapSort:
    """Representra el HeapSort que ordenarà la llista."""
    def __init__(self, llista: list) -> None:
        # Comprovar que el valor passat sigui una llista.
        if not isinstance(llista, list):
            raise NoEsLLista

        # Comprovar que tots els valors de la llista siguin nombres.
        for x in llista:
            if not isinstance(x, int):
                raise ValorsNoNumerics
        
        # Comprovar que la llista no sigui buida.
        if len(llista) == 0:
            raise LlistaBuida

        self.llista = llista

    def ordenar(self) -> None:
        """Ordena la llista."""
        num_nodes = len(self.llista)

        print(Fore.CYAN + "Buscant nombre més gran.")

        # Trobar i pujar el número més gran.
        for i in range(num_nodes // 2, -1, -1):
            self.heapify(num_nodes, i)

        print(Fore.CYAN + "Fi del primer bucle.")
        print(Fore.CYAN + f"El número més gran es {self.llista[0]}.")

        # Pujar els nombres restants i ordenar-los.
        for i in range(num_nodes - 1, 0, -1):
            self.intercanvi(i, 0)
            self.heapify(i, 0)
        
        print(Fore.CYAN + "Llista ordenada.")

    def heapify(self, num_nodes: int, node_actual: int) -> None:
        """'heapify' es la funció cridada per 'odernar' que fa la comparació
        del node passat per parametre i dels seus nodes inferiors.
        
        Args:
            num_nodes (int): Nombre de nodes restants a l'arbre.
            node_actual (int): Node a comparar.
        """
        print(Fore.GREEN + f"Inici del Heapify pel node {node_actual}.")
        major = node_actual
        esquerra = 2 * node_actual + 1
        dreta = 2 * node_actual + 2

        # En cas que hi hagi un node inferior a l'esquerra, comprovar si aquest node
        # és més petit que el node actual.
        if (
            esquerra < num_nodes
            and self.llista[node_actual] < self.llista[esquerra]
        ):
            print(Fore.MAGENTA + "El node esquerre passa a ser el major.")
            major = esquerra

        # En cas que hi hagi un node inferior a la dreta, comprovar si el node major
        # és més petit que el node de la dreta.
        if dreta < num_nodes and self.llista[major] < self.llista[dreta]:
            print(Fore.MAGENTA + "El node dret passa a ser el major.")
            major = dreta

        # Si el node actual ha deixat de ser el més gran, es canvien i es
        # continua fent Heapify recursivament.
        if major != node_actual:
            print("Canviant el node per el número major.")
            self.intercanvi(node_actual, major)
            self.heapify(num_nodes, major)
        
        else:
            print("El número major no ha canviat. No es fa res.")
        
        print(Fore.GREEN + "Final del Heapify.")

    def intercanvi(self, x: int, y: int) -> None:
        """Fa un intercanvi entre dos índex de la llista.

        Args:
            x (int): Primer índex.
            y (int): Segón índex.
        """
        print(
            f"Intercanvi del node {x} (valor: {self.llista[x]}) amb"
            + f" el node {y} (valor: {self.llista[y]})"
        )
        self.llista[x], self.llista[y] = self.llista[y], self.llista[x]

File: src/main/java/com/heapsort/HeapSort.java
/**
 * HeapSort implementat a Java.
 * 
 * @author Sergi Serra, Yasmine Fadiss, Jessica Martinez, Arnau Vilalta, Iván Tovar, Pol Sanchez, Maximiliano Godoy
 * @version 1.0
 */
package main.java.com.heapsort;

import java.util.ArrayList;

import main.java.com.heapsort.exceptions.LlistaBuida;
import main.java.com.heapsort.exceptions.NoEsLlista;
import main.java.com.heapsort.exceptions.ValorsNoNumerics;

/**
 * Representra el HeapSort que ordenarà la llista.
 * 
 * @author Sergi Serra, Yasmine Fadiss, Jessica Martinez, Arnau Vilalta, Iván Tovar, Pol Sanchez, Maximiliano Godoy
 */
public class HeapSort {
	private ArrayList<Integer> numeros;

	/**
	 * Crea una objecte HeapSort llest per ordenar la llista.
	 * 
	 * @param numeros Llista de números.
	 * @throws NoEsLlista Si el parametre passat no es una llista de tipus ArrayList.
	 * @throws ValorsNoNumerics Si la llista té valors no numérics.
	 * @throws LlistaBuida Si la llista es buida.
	 */
	public HeapSort(ArrayList<Integer> numeros) throws NoEsLlista, ValorsNoNumerics, LlistaBuida {
		// Comprovar que el valor passat sigui una llista.
		if (!(numeros instanceof ArrayList)) {
			throw new NoEsLlista();
		}

		// Comprovar que tots els valors de la llista siguin enters.
		for (Integer x : numeros) {
			if (!(x instanceof Integer)) {
				throw new ValorsNoNumerics();
			}
		}

		// Comprovar que la llista no estigui buida.
		if (numeros.isEmpty()) {
			throw new LlistaBuida();
		}

		this.numeros = numeros;
	}

	/**
	 * Ordena la llista.
	 */
	public void ordenar() {
		int n = numeros.size();

		// Trobar i pujar el número més gran.
		for (int i = n / 2 - 1; i >= 0; i--) {
			heapify(numeros, n, i);
		}

		System.out.println("Fi del primer bucle.");
		System.out.println("El número més gran es " + numeros.get(0) + ".");

		// Pujar els nombres restants i ordenar-los.
		for (int i = n - 1; i >= 0; i--) {
			intercanvi(0, i);
			heapify(numeros, i, 0);
		}
	}

	/**
	 * 'heapify' és el métode cridat per 'ordenar' que fa la comparació
	 * del node passat per paràmetre i dels seus nodes inferiors (fills).
	 * 
	 * @param numeros Llista de números.
	 * @param num_nodes Nombre de nodes restants a l'arbre.
	 * @param node_actual Node a comparar.
	 */
	void heapify(ArrayList<Integer> numeros, int num_nodes, int node_actual) {
		// 
		int major = node_actual;
		int esquerra = 2 * node_actual + 1;
		int dreta = 2 * node_actual + 2;

		// En cas que hi hagi un node inferior a l'esquerra, comprovar si el node
		// esquerra és més petit que el node actual.
		if (esquerra < num_nodes && numeros.get(esquerra) > numeros.get(major)) {
			System.out.println("El node esquerra passa a ser el major.");
			major = esquerra;
		}

		// En cas que hi hagi un node inferior a la dreta, comprovar si el node major
		// és més petit que el node dreta.
		if (dreta < num_nodes && numeros.get(dreta) > numeros.get(major)) {
			System.out.println("El node dret passa a ser el major.");
			major = dreta;			
		}

		// Si el node_actual ha deixat de ser el més gran, es canvien i es
		// continua fent Heapify recursivament.
		if (major != node_actual) {
			System.out.println("Canviant el node per el número major.");
			intercanvi(node_actual, major);
			heapify(numeros, num_nodes, major);

		} else {
			System.out.println("El número major no ha canviat. No es fa res.");
		}

		System.out.println("Final del Heapify");
	}

	/**
	 * Fa un intercanvi entre dos índex de la llista.
	 * 
	 * @param x Primer índex.
	 * @param y Segon índex.
	 */
	void intercanvi(int x, int y) {
		int swap = numeros.get(x);
		numeros.set(x, numeros.get(y));
		numeros.set(y, swap);
	}

	/**
	 * Mostra la llista.
	 * 
	 * @param numeros Llista de números.
	 */
	static void mostra_llista(ArrayList<Integer> numeros) {
		for (int i = 0; i < numeros.size(); i++) {
			System.out.print(numeros.get(i) + " ");
		}
		System.out.println();
	}
}
